haystack/components/rankers/sentence_transformers_diversity.py

```diff
@@ -218,17 +234,88 @@ def _greedy_diversity_order(self, query: str, documents: List[Document]) -> List
+            doc_embeddings /= torch.norm(doc_embeddings, p=2, dim=-1).unsqueeze(-1)
+            query_embedding /= torch.norm(query_embedding, p=2, dim=-1).unsqueeze(-1)
+        return doc_embeddings, query_embedding
+
+    def _maximum_margin_relevance(
+        self, query: str, documents: List[Document], lambda_threshold: float = 0.5
+    ) -> List[Document]:
+        """
+        Orders the given list of documents according to the Maximum Margin Relevance (MMR) scores.
+
+        MMR scores are calculated for each document based on their relevance to the query and diversity from already
+        selected documents.
+
+        The algorithm iteratively selects documents based on their MMR scores, balancing between relevance to the query
+        and diversity from already selected documents. The 'lambda_threshold' controls the trade-off between relevance
+        and diversity.
+
+        A closer value to 0, favors diversity, while a closer value to 1, favors relevance to the query.
+
+        See : "The Use of MMR, Diversity-Based Reranking for Reordering Documents and Producing Summaries"
+               https://www.cs.cmu.edu/~jgc/publication/The_Use_MMR_Diversity_Based_LTMIR_1998.pdf
+        """
+
+        texts_to_embed = self._prepare_texts_to_embed(documents)
+
+        doc_embeddings, query_embedding = self._embed_and_normalize(query, texts_to_embed)
+
+        n = len(documents)
+        selected: List[int] = []
+
+        # compute MMR score for each document
+        mmr_scores = []
+        for i in range(n):
+            relevance_score = query_embedding @ doc_embeddings[i].T
+            # diversity score: max similarity score between current document and all the current selected documents
+            if len(selected) > 0:
+                diversity_score = max(doc_embeddings[i] @ doc_embeddings[j] for j in selected)
+            else:
+                diversity_score = 0
+            mmr_score = lambda_threshold * relevance_score - (1 - lambda_threshold) * diversity_score
+            mmr_scores.append(mmr_score)
+            selected.append(i)
+
+        # rank documents based on MMR score
+        ranked_docs = [doc for _, doc in sorted(zip(mmr_scores, documents), reverse=True)]
+
+        return ranked_docs
+
+    @staticmethod
+    def _check_lambda_threshold(lambda_threshold, strategy):
+        if strategy == "maximum_margin_relevance" and (lambda_threshold < 0 or lambda_threshold > 1):
+            raise ValueError(f"lambda_threshold must be between 0 and 1, but got {lambda_threshold}.")
+
     @component.output_types(documents=List[Document])
-    def run(self, query: str, documents: List[Document], top_k: Optional[int] = None):
+    def run(  # pylint: disable=too-many-positional-arguments
+        self,
+        query: str,
+        documents: List[Document],
+        top_k: Optional[int] = None,
+        strategy: Literal["greedy_diversity_order", "maximum_margin_relevance"] = "greedy_diversity_order",
```

Comment:
Why does it make sense to expose this as a runtime parameter?

Reply:
Why not?

Reply:
But what's the rational behind it? having it in run time one can easily experiment which strategy gives the best results without re-initialising the component